upbc/protoc-gen-upb.cc

```diff
@@ -418,294 +418,362 @@ void GenerateExtensionInHeader(const protobuf::FieldDescriptor* ext,
+      "UPB_INLINE $0* $1_resize_$2($1 *msg, size_t len, "
+      "upb_Arena *arena) {\n"
+      "  return ($0*)_upb_Array_Resize_accessor2(msg, $3, len, $4, "
+      "arena);\n"
+      "}\n",
+      CType(field), msg_name, field->name(),
+      GetSizeInit(layout.GetFieldOffset(field)), SizeLg2(field));
+  if (field->cpp_type() == protobuf::FieldDescriptor::CPPTYPE_MESSAGE) {
     output(
-        "  $0_NOT_SET = 0\n"
-        "} $0_oneofcases;\n",
-        fullname);
+        "UPB_INLINE struct $0* $1_add_$2($1 *msg, upb_Arena *arena) {\n"
+        "  struct $0* sub = (struct $0*)_upb_Message_New(&$3, arena);\n"
+        "  bool ok = _upb_Array_Append_accessor2(\n"
+        "      msg, $4, $5, &sub, arena);\n"
+        "  if (!ok) return NULL;\n"
+        "  return sub;\n"
+        "}\n",
+        MessageName(field->message_type()), msg_name, field->name(),
+        MessageInit(field->message_type()),
+        GetSizeInit(layout.GetFieldOffset(field)), SizeLg2(field));
+  } else {
     output(
-        "UPB_INLINE $0_oneofcases $1_$2_case(const $1* msg) { "
-        "return ($0_oneofcases)*UPB_PTR_AT(msg, $3, int32_t); }\n"
-        "\n",
-        fullname, msg_name, oneof->name(),
-        GetSizeInit(layout.GetOneofCaseOffset(oneof)));
+        "UPB_INLINE bool $1_add_$2($1 *msg, $0 val, upb_Arena *arena) {\n"
+        "  return _upb_Array_Append_accessor2(msg, $3, $4, &val,\n"
+        "      arena);\n"
+        "}\n",
+        CType(field), msg_name, field->name(),
+        GetSizeInit(layout.GetFieldOffset(field)), SizeLg2(field));
   }
+}
 
-  // Generate const methods.
+void GenerateNonRepeatedSetters(const protobuf::FieldDescriptor* field,
+                                const MessageLayout& layout,
+                                absl::string_view msg_name, Output& output) {
+  if (field == field->containing_type()->map_key()) {
+    // Key cannot be mutated.
+    return;
+  }
 
-  for (auto field : FieldNumberOrder(message)) {
-    // Generate hazzer (if any).
-    if (layout.HasHasbit(field)) {
-      output(
-          "UPB_INLINE bool $0_has_$1(const $0 *msg) { "
-          "return _upb_hasbit(msg, $2); }\n",
-          msg_name, field->name(), layout.GetHasbitIndex(field));
-    } else if (field->real_containing_oneof()) {
-      output(
-          "UPB_INLINE bool $0_has_$1(const $0 *msg) { "
-          "return _upb_getoneofcase(msg, $2) == $3; }\n",
-          msg_name, field->name(),
-          GetSizeInit(
-              layout.GetOneofCaseOffset(field->real_containing_oneof())),
-          field->number());
-    } else if (field->message_type()) {
-      output(
-          "UPB_INLINE bool $0_has_$1(const $0 *msg) { "
-          "return _upb_has_submsg_nohasbit(msg, $2); }\n",
-          msg_name, field->name(), GetSizeInit(layout.GetFieldOffset(field)));
-    }
+  // The common function signature for all setters.  Varying
+  // implementations follow.
+  output("UPB_INLINE void $0_set_$1($0 *msg, $2 value) {\n", msg_name,
```

Comment:
I think (not 100% sure) this it will make things more readable if you repeat the header so that each snippet of code generation is locally complete